### PR TITLE
Support for AlwaysRelevent object references.

### PR DESCRIPTION
**Pull request by Vatyx**
#### Description
Object references can now be annotated with `AlwaysRelevant`. This will make the Actor have interest in the other Object. This is only supported for Actors and not subobjects, at the moment.
#### Documentation
Will be added later since I think we need to refine this API a bit.

#### Primary reviewers
@danielimprobable @m-samiec

### CODE REVIEW

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -1047,7 +1051,7 @@ void USpatialNetDriver::TickFlush(float DeltaTime)
 #if USE_SERVER_PERF_COUNTERS
 	double ServerReplicateActorsTimeMs = 0.0f;
 #endif // USE_SERVER_PERF_COUNTERS
-	if (IsServer() && ClientConnections.Num() > 0)
+	if (IsServer() && ClientConnections.Num() > 0 && Connection->IsConnected())
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Drive-by?

File: SpatialGDK/Source/Private/Interop/SpatialTypebindingManager.cpp [hunk 2]
```diff
@@ -117,6 +117,20 @@ void USpatialTypebindingManager::CreateTypebindings()
 					Info.HandoverProperties.Add(HandoverInfo);
 				}
 			}
+
+			if (Property->PropertyFlags & CPF_AlwaysInterested)
+			{
+				for (int32 ArrayIdx = 0; ArrayIdx < PropertyIt->ArrayDim; ++ArrayIdx)
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialTypebindingManager.cpp]: Duplicate code, template it?

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 3]
```diff
@@ -141,6 +143,7 @@ Worker_RequestId USpatialSender::CreateEntity(USpatialActorChannel* Channel)
 	ComponentDatas.Add(improbable::Persistence().CreatePersistenceData());
 	ComponentDatas.Add(improbable::Rotation(Actor->GetActorRotation()).CreateRotationData());
 	ComponentDatas.Add(improbable::UnrealMetadata({}, ClientWorkerAttribute, Class->GetPathName()).CreateUnrealMetadataData());
+	ComponentDatas.Add(improbable::Interest().CreateInterestData());
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Need this for Startup Actors also?

File: SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp (hunk 1 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/EngineClasses/SpatialNetDriver.cpp]: Yes. I'm not sure how this ever worked without that condition.

File: SpatialGDK/Source/Private/Utils/ComponentFactory.cpp [hunk 4]
```diff
@@ -286,6 +294,8 @@ TArray<Worker_ComponentData> ComponentFactory::CreateComponentDatas(UObject* Obj
 		ComponentDatas.Add(CreateHandoverComponentData(Info->SchemaComponents[SCHEMA_Handover], Object, Info, HandoverChangeState));
 	}
 
+	ComponentDatas.Add(CreateInterestComponentData(Object, Info));
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Utils/ComponentFactory.cpp]: Don't we only support Actor interest currently?

File: SpatialGDK/Source/Private/Utils/ComponentFactory.cpp [hunk 5]
```diff
@@ -413,4 +430,70 @@ Worker_ComponentUpdate ComponentFactory::CreateHandoverComponentUpdate(Worker_Co
...
+}
+
+Worker_ComponentUpdate ComponentFactory::CreateInterestComponentUpdate(UObject* Object, FClassInfo* Info)
+{
+	return CreateInterestComponent(Object, Info).CreateInterestUpdate();
+}
+
+improbable::Interest ComponentFactory::CreateInterestComponent(UObject* Object, FClassInfo* Info)
+{
+	//Create a new component interest containing a query for every interested Object
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Utils/ComponentFactory.cpp]:
```suggestion
	// Create a new component interest containing a query for every interested Object
```

File: SpatialGDK/Source/Private/Utils/ComponentFactory.cpp [hunk 6]
```diff
@@ -413,4 +430,70 @@ Worker_ComponentUpdate ComponentFactory::CreateHandoverComponentUpdate(Worker_Co
...
+	return Interest;
+}
+
+void ComponentFactory::AddObjectToComponentInterest(UObject* Object, UObjectPropertyBase* Property, uint8* Data, improbable::ComponentInterest& ComponentInterest)
+{
+	UObject* ObjectOfInterest = Property->GetObjectPropertyValue(Data);
+
+	if (ObjectOfInterest == nullptr)
+	{
+		return;
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Utils/ComponentFactory.cpp]: When would this happen?

File: SpatialGDK/Source/Public/Schema/StandardLibrary.h [hunk 7]
```diff
@@ -44,6 +47,29 @@ struct Coordinates
 	}
 };
 
+using EdgeLength = Coordinates;
```
Comment (m-samiec) [inline SpatialGDK/Source/Public/Schema/StandardLibrary.h]: Is this used?

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp (hunk 3 above)
Comment (m-samiec) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Ahh see it below.

File: SpatialGDK/Source/Private/Interop/SpatialTypebindingManager.cpp (hunk 2 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialTypebindingManager.cpp]: I removed properties in the FInterestPropertInfo struct due to parts being unused so it's not duplicated any more really.

File: SpatialGDK/Source/Private/Utils/ComponentFactory.cpp (hunk 4 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Utils/ComponentFactory.cpp]: Yup, good catch.

File: SpatialGDK/Source/Private/Utils/ComponentFactory.cpp (hunk 6 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Utils/ComponentFactory.cpp]: When a property goes from pointing to an object to being explicitly set to nullptr.

File: SpatialGDK/Source/Public/Schema/StandardLibrary.h (hunk 7 above)
Comment (Vatyx) [reply SpatialGDK/Source/Public/Schema/StandardLibrary.h]: Yes in Interest.h. I've moved it.

File: SpatialGDK/Source/Public/Utils/ComponentFactory.h [hunk 8]
```diff
@@ -19,6 +20,7 @@ class UProperty;
 enum EReplicatedPropertyGroup : uint32;
 
 using FUnresolvedObjectsMap = TMap<Schema_FieldId, TSet<const UObject*>>;
+using FResolvedActorProxySet = TSet<const UObject*>;
```
[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/Public/Utils/ComponentFactory.h]: Nope, removed.

File: SpatialGDK/Source/Public/Schema/UnrealObjectRef.h [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+	T* operator->()
+	{
+		return Value.Get();
+	}
+
+private:
+	TUniquePtr<T> Value;
+};
+
+}
```
Comment (m-samiec) [inline SpatialGDK/Source/Public/Schema/UnrealObjectRef.h]: Slightly tangental, but shouldn't everything in this file be in the `improbable` namespace?
